src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineHelper.cs

```diff
@@ -306,38 +246,62 @@ public static ImmutableArray<DocumentSymbolUIItem> GetDocumentSymbolUIItems(Immu
         }
 
         /// <summary>
-        /// Updates the IsExpanded property for the Document Symbol UI representation based on the given Expansion Option. The parameter
+        /// Updates the IsExpanded property for the Document Symbol ViewModel based on the given Expansion Option. The parameter
         /// <param name="currentDocumentSymbolItems"/> is used to reference the current node expansion in the view.
         /// </summary>
-        public static void SetIsExpanded(
-            ImmutableArray<DocumentSymbolUIItem> documentSymbolItems,
-            IEnumerable<DocumentSymbolUIItem> currentDocumentSymbolItems,
+        public static void SetIsExpandedOnNewItems(
+            ImmutableArray<DocumentSymbolDataViewModel> newDocumentSymbolItems,
+            ImmutableArray<DocumentSymbolDataViewModel> currentDocumentSymbolItems)
+        {
+            using var _ = PooledHashSet<DocumentSymbolDataViewModel>.GetInstance(out var hashSet);
+            hashSet.AddRange(newDocumentSymbolItems);
+
+            foreach (var item in currentDocumentSymbolItems)
+            {
+                if (!hashSet.TryGetValue(item, out var newItem))
+                {
+                    continue;
+                }
+
+                // Setting a boolean property on this View Model is allowed to happen on any thread.
+                newItem.IsExpanded = item.IsExpanded;
+                SetIsExpandedOnNewItems(newItem.Children, item.Children);
+            }
+        }
+
+        public static void SetExpansionOption(
+            ImmutableArray<DocumentSymbolDataViewModel> currentDocumentSymbolItems,
             ExpansionOption expansionOption)
```

Comment:
i presume something above this doesn't bother calling this if the expansionOption is NoChange or CurrentExpansion?  If not, it should, so we don't waste time warlking. 

Can you then just make this a boolean for expand/collapse?